### ISSUE TITLE
DTMF Buttons sometimes not shown

### ISSUE THREAD

**Issue by dl1nux** [f5vmr/SVXLink-Dash-V2#5]
The dashboard code has an issue what has been inherited from the original versions. The dashboard distinguishes between "local" and "remote" addresses when viewing the dashboard. "Local" IP addresses will show the DTMF buttons and an input field. Viewing from remote adresses will not show them.

The definition of the "local" adresses is a little bit stupid (not your fault Chris) and should be corrected in .../include/buttons.php

Wrong definition! :
$net1= cidr_match($ip,"192.168.0.0/8");
$net2= cidr_match($ip,"192.175.43.91/8");
$net3= cidr_match($ip,"127.0.0.0/8");
$net4= cidr_match($ip,"192.168.1.0/8");
$net5 = cidr_match($ip, "192.168.1.254/8");

Correct definition of private adresses:
$net1= cidr_match($ip,"192.168.0.0/16");
$net2= cidr_match($ip,"10.0.0.0/8");
$net3= cidr_match($ip,"127.0.0.0/8");
$net4= cidr_match($ip,"172.16.0.0/12");
$net5= cidr_match($ip,"169.254.0.0/16");

After that change all can see the buttons when viewing the dashboard locally. It can be adapted when using special networks.

73 Attila DL1NUX

**Comment by dl1nux** [comment]
And as addition, you are writing in the README:

"Below the footer menu you will see a dtmf control facility. The D buttons are not yet programmed, and will need to correspond with the Macros in svxlink.conf so you will need to self-edit the dtmf.php file in your dashboard configuration and the macros in svxlink.conf."

It's only neccessary to edit config.inc.php to personalize the DTMF buttons, for example:

define("KEY1", array(' Hassberge ','*9197475#','green'));
define("KEY2", array(' Franken ','*9126284#','orange'));
define("KEY3", array(' Bayern ','*912628#','orange'));
define("KEY4", array(' RAMSES ','*9165075#','orange'));
define("KEY5", array(' Papagei', '1#','purple'));
define("KEY6", array(' Echolink ','2#','purple'));
define("KEY7", array(' EL DB0UC ','2#96450#','purple'));
define("KEY8", array(' EL Echotest ','2#9999#','blue'));
define("KEY9", array(' Trennen ','##','blue'));
define("KEY10", array(' Status ','9*#','red'));

No need to edit macros or dtmf.php.

**Comment by f5vmr** [comment]
To be honest, these should be only for the local user. I have not configured any of mine for anything but local interaction. I expect there will be likely some people playing with this if the dashboard is exposed on a global address.
I’m still working on a better edit scheme on another repo at the moment. I’ll post that when It’s complete.
The DTMF part was made by SP2ONG and his associates.

Chris

**Comment by dl1nux** [comment]
Yes, the buttons are only for local usage. But they can't be used if the local network does not match to the hard coded networks, which are mostly incorrect. Here we had several users who couldnt use the buttons (locally!) because they have other network addresses. It's just a hint, because most people would not know why they cant see the buttons.

**Comment by DC9VQ** [comment]
any hints how to activate DTMF with the buttons/input field ? I guess I have to change anything in den svxlink.conf but It doesn't work so far.

resolved with this great manual chapter 6.4 but only in German 
https://www.bravo37.de/einrichtung-rf-guru-svxlink-hotspot/

**Comment by dl1nux** [comment]
> any hints how to activate DTMF with the buttons/input field ? I guess I have to change anything in den svxlink.conf but It doesn't work so far.
> 
> resolved with this great manual chapter 6.4 but only in German https://www.bravo37.de/einrichtung-rf-guru-svxlink-hotspot/

Thanks for linking, this is my article I posted on or clubs website, because we have several hotspot users (RF Guru).
The old version is here, it included more repair hints for the dashboard (there were several bugs):
https://www.bravo37.de/einrichtung-rf-guru-image-januar-2024/

**Comment by f5vmr** [comment]
Hi Everybody,

I strongly suspect this to be an issue of the ownership of the Webserver being unable to send the correct command to the appropriate file.

As I understand it in SimplexLogic or RepeaterLogic the following line has to be enabled. In my case it is simplex

DTMF_CTRL_PTY = /dev/shm/simplex_dtmf_ctrl

When the service is run /dev/shm  will be allocated a pseudo terminal in which to run.

The dashboard wants to send DTMF Commands via /tmp/dtmf_svx, so this location has to be changed to a symlink pointing to the pseudo terminal that svxlink has chosen.

So we type ls -l /dev/shm - Depending on the result this probably will be something similar to /dev/pts/0 or 1 or 2. Mine is /dev/pts/1.

The last step is to create the file - sudo touch /tmp/dtmf_svx 
sudo chown svxlink:svxlink /tmp/dtmf_svx
sudo chmod 666 /tmp/dtmf_svx

Then we issue the symlink - sudo ln -sf /dev/pts/1 /tmp/dtmf_svx

When we examine the symlink ls -l /tmp/dtmf/dtmf_svx is should now point to the /dev/pts/1 with ownership svxlink:tty

However I have yet to make the echo command in PHP perform correctly, even with special permissions granted to the dashboard owner svxlink in sudoers.d/svxlink.

However when I log into the terminal and sudo -I to become user root, and type echo “91235#” > /tmp/dtmf the command is fully and correctly interpreted.

The operation for the DTMF band commands in the PHP file - /var/www/html/include/buttons.php where you will find the operations button by button.

This is code for the first button with debugging lines included. 

if(array_key_exists('button1', $_POST)) {
        $exec= "echo '" . KEY1[1] . "' > /tmp/dtmf_svx";
        debug_log("Button 1 Executed");
            exec($exec,$output);
        debug_log("output: " . print_r($output,true));
            echo "<meta http-equiv='refresh' content='0'>";
        }

However the debugging fails indicating that the exec command has no permission to run. 

I am still working on this, and hope to resolve it soon.

Chris G4NAB

**Comment by f5vmr** [comment]
Incidentally https://portal.svxlink.uk:8443/downloads will provide an image that will work in most hotspots and repeaters, providing you know the correct gpio functions.
It also installs the WiFi hotspot, and the Svxlink Dashboard (UK Model with editing of the .conf files).
It works with SA818 models as well as sound card models with wm8960 and CM108/119.

Chris

**Comment by f5vmr** [comment]
The latest dashboard material fixes completely most of the issues on the dashboard
go to /var/www/html and perform sudo git pull - if you receive errors then remove include/config.inc.php making note of your password if complicated, and repeater the sudo git pull. Afterwhich sudo ./upgrade.sh to make chenges. Note - there is an instructions.txt file - Follow the instructions to make your Rx Monitor work as it should. You should already have a loopback card installed - check aplay -l. 73
